---
title: "During: OF vs OC"
output: 
  html_notebook:
    theme: flatly
    toc: true
    toc_float: true
    code_folding: hide
---

```{r, message=FALSE, message=FALSE, warning=FALSE}
library(tidyverse)
library(biomaRt)
library(ggrepel)
library(clusterProfiler)
library(enrichplot)
# Parallel
library(BiocParallel)
register(MulticoreParam(6))

load('../data/microarray_NGS_objects.Rdata')
load('../data/top_tables.Rdata')
sva_counts <- read_tsv('../data/sva_counts.tsv.gz')

sample_meta_D <- sample_meta %>% filter(Sample %in% colnames(sva_counts)) %>%
  dplyr::select(Sample:Section, Layout:Fusion) %>%
  mutate(S2 = case_when(Section == 'OF' ~ 'OF', TRUE ~ 'OC')) %>%
  unique()

box_maker <- function(table, genes, section = c('OF','OC'), type = 'temporal'){
  if ('matrix' %in% class(table)){
    table <- table %>%
      as_tibble(rownames = 'Gene')
  }
  if (type == 'temporal'){
    plot <- table %>% 
      pivot_longer(-Gene, names_to = 'Sample', values_to = 'Expression') %>%
      mutate(Sample = gsub('_.*|.CEL.*','',Sample)) %>%
      left_join(sample_meta_D) %>%
      mutate(S2 = case_when(Section == 'OF' ~ 'OF',TRUE ~ 'OC')) %>%
      filter(Gene %in% genes, S2 %in% section) %>%
      #filter(Gene %in% row.names(top.table_OF_AD %>% head(10))) %>%
      mutate(Fusion = factor(Fusion, levels = c('Before','During','After'))) %>%
      mutate(OrgTech = paste(Organism, Technology, sep = '_')) %>% 
      ggplot(aes(x=Fusion, y=Expression, color = Organism, shape = Technology)) +
      # geom_boxplot(aes(group = Fusion), color = 'Black', outlier.colour = NA) +
      # geom_boxplot(aes(group = interaction(Organism,Fusion)), outlier.colour = NA) +
      ggbeeswarm::geom_quasirandom(size = 3, alpha = 0.7) +
      cowplot::theme_cowplot() +
      facet_grid(~Gene + S2, scales = 'free_y') +
      ggsci::scale_color_aaas() +
      ylab('log2 (corrected counts)') +
      stat_summary(fun=mean, geom = 'line', aes(group = OrgTech, color = Organism)) }
  else {
    plot <- table %>% 
      pivot_longer(-Gene, names_to = 'Sample', values_to = 'Expression') %>%
      mutate(Sample = gsub('_.*|.CEL.*','',Sample)) %>%
      left_join(sample_meta_D) %>%
      mutate(S2 = case_when(Section == 'OF' ~ 'OF',TRUE ~ 'OC')) %>%
      filter(Gene %in% genes, S2 %in% section) %>%
      mutate(Fusion = factor(Fusion, levels = c('Before','During','After'))) %>%
      filter(Fusion == 'During') %>% 
      mutate(OrgTech = paste(Organism, Technology, sep = '_')) %>% 
      ggplot(aes(x=S2, y=Expression, color = Organism, shape = Technology)) +
      # geom_boxplot(aes(group = Fusion), color = 'Black', outlier.colour = NA) +
      # geom_boxplot(aes(group = interaction(Organism,Fusion)), outlier.colour = NA) +
      ggbeeswarm::geom_quasirandom(size = 3, alpha = 0.7) +
      cowplot::theme_cowplot() +
      ggsci::scale_color_aaas() +
      ylab('log2 (corrected counts)') +
      xlab('Section') +
      stat_summary(fun=mean, geom = 'line', aes(group = OrgTech, color = Organism)) + facet_wrap(~Gene)
  }
  plot
}

volcano_maker <- function(df, 
                          title="Volcano Plot", 
                          pvalue='P.Value', 
                          padj='adj.P.Val', 
                          logFC='logFC', 
                          gene_list = ''){
  df$pvalue <- df[,pvalue]
  df$log2FoldChange <- df[,logFC]
  df$padj <- df[,padj]
  df$Gene <- row.names(df)
  df <- df[!is.na(df$pvalue),]
  print(dim(df))
  
  df <- df %>% mutate(Class = case_when(padj < 0.05 & abs(logFC) > 1~ "FDR < 0.05 & logFC > 1",
                                        padj < 0.1 & abs(logFC) > 1 ~ 'FDR < 0.1 & logFC > 1',
                                        TRUE ~ 'Not significant'))
  df$GeneT <- df$Gene
  if (gene_list == ''){
    gene_list <- df %>% filter(padj < 0.05) %>% pull(Gene) %>% head(10)
  }
  df$Gene[!df$Gene %in% gene_list] <- ''
  
  plot <- ggplot(data=df,aes(label=Gene, x = log2FoldChange, y = -log10(pvalue))) +
    geom_point(aes(colour=Class)) +
    scale_colour_manual(values=c("darkred", "red", "grey")) +
    cowplot::theme_cowplot() +
    geom_vline(aes(xintercept=-1),linetype="dotted") +
    geom_vline(aes(xintercept=1),linetype="dotted") +
    geom_vline(aes(xintercept=-2),linetype="dotted") +
    geom_vline(aes(xintercept=2),linetype="dotted") +
    geom_label_repel(max.overlaps = 100) +
    xlab('logFC') + ylab('-log10(p value)') +
    ggtitle(title) + cowplot::theme_cowplot()
  
  plot
}
```

# During vs Before (OF)

2021-12-06

**Positive means higher expression in the OF relative to the OCn** This test is designed to identify which genes are changing between the OF and OC in the "during" stage.

## Volcano

```{r, fig.width=5, fig.height=2}
volcano_maker(top.table_During, title = 'During: OF vs OC',
              
              gene_list = c(top.table_During %>% filter(logFC > 0) %>% head(12) %>% row.names(),
                            top.table_During %>% filter(logFC < 0) %>% head(12) %>% row.names()))
```

## Diff Table

All genes with an FDR < 0.2. Not very many. There are a limited number of samples *before* the fusion begins (for the OF - no human) compared to the other time points. 

```{r}
top.table_During %>% as_tibble(rownames = 'Gene') %>% filter(adj.P.Val < 0.1) %>%  DT::datatable()
```

## Expression of top 6 genes (by FDR) across section and stage

Colored by organism. Each line is drawn for organism / technology (remember, mouse has both microarray and RNA-seq).


```{r, fig.width=12, fig.height=6}
box_maker(sva_counts, 
          genes = top.table_During %>% 
            as_tibble(rownames = 'Gene') %>% 
            filter(adj.P.Val < 0.05) %>% head(24) %>% pull(Gene),
          section = c('OF','OC'), type = 'bloop')


```

# Enrichment Analysis (GO, GSEA) Before -> During

##  GSEA

GSEA uses a *ranked list* of genes by logFC. So the p values are not used in this situation. The *order* is. So the GSEA is useful in situations where there are very few differentially expressed genes. 

Positive terms (higher in OF) relate to collagen / extracellular matrix, basement membrane. Negative terms (higher in OC) relate to cell cycle. 

### Dotplot

```{r, fig.width=7, fig.height=6}
all_genes <- bitr(top.table_During %>% as_tibble(rownames = 'Gene') %>% pull(Gene), fromType="SYMBOL", toType="ENTREZID", OrgDb="org.Hs.eg.db")
all_genes <- all_genes %>% left_join(top.table_During %>% as_tibble(rownames = 'SYMBOL'), by = c('SYMBOL'))

logFC <- all_genes$logFC
names(logFC) <- all_genes$ENTREZID
logFC <- na.omit(logFC)

logFC = sort(logFC, decreasing = TRUE)

gse <- gseGO(geneList=logFC,
             ont ="ALL",
             keyType = "ENTREZID",
             pvalueCutoff = 0.05,
             OrgDb = org.Hs.eg.db,
             pAdjustMethod = "BH",
             eps = 0)
gse <- setReadable(gse, OrgDb = org.Hs.eg.db)
# change sort logic
gseF <- gse
gseF@result <- rbind(gseF@result %>% arrange(NES) %>% head(20),
                     gseF@result %>% arrange(NES) %>% tail(20) %>% arrange(-NES)
)
dotplot(gse, showCategory=15, split=".sign") + facet_grid(.~.sign) + cowplot::theme_cowplot()
```

### Related terms

So we can see "chunks" of terms that go together. We see several terms related to the structural components of the cell. As well as a set of cell cycle terms. 
```{r, fig.width=5, fig.height=4}
gsePT <- pairwise_termsim(gse)
emapplot(gsePT)
```

### Table

So you can see the genes in the ontology term. The genes get "included" as enriched if GSEA deems them to be ranked unusually high.
```{r}
gse@result %>% as_tibble() %>% arrange(-abs(NES)) %>% filter(p.adjust < 0.05) %>%  DT::datatable()
```

## GO Enrichment 

GO enrichment uses a *cutoff* between differentially expressed genes (FDR < 0.1 in this case) and everything else. 

Loads of stuff relating to visual function and development. 
```{r, fig.width=7}



diff_genes <- top.table_During %>% as_tibble(rownames = 'Gene') %>% filter(adj.P.Val < 0.1) %>% 
  mutate(Group = case_when(logFC > 0 ~  'Higher in OF', TRUE ~ 'Higher in OC'))
eg_diff_genes <- bitr(diff_genes$Gene, fromType="SYMBOL", toType="ENTREZID", OrgDb="org.Hs.eg.db")
eg_diff_genes <- diff_genes %>% left_join(., eg_diff_genes, by = c('Gene' = 'SYMBOL'))
eg_universe = bitr(top.table_During %>% as_tibble(rownames = 'Gene') %>% pull(Gene), fromType="SYMBOL", toType="ENTREZID", OrgDb="org.Hs.eg.db")

eg_diff_gene_list <- eg_diff_genes$logFC
names(eg_diff_gene_list) <- eg_diff_genes$ENTREZID

egoOF_OC <- enrichGO(gene          = eg_diff_genes$ENTREZID,
                     universe      = eg_universe$ENTREZID,
                     OrgDb         = org.Hs.eg.db,
                     ont           = "all",
                     readable      = TRUE)



p1 <- dotplot(egoOF_OC, showCategory=20) + ggtitle("Dotplot for GO, OF vs OC")
p1


```
### Table

So you can see the genes in the ontology term.
```{r}
egoOF_OC@result %>% as_tibble() %>%  filter(p.adjust < 0.05) %>%  DT::datatable()
```

### CNET Plot

Relationships between related GO terms with shared genes. Yellow means more expressed in the OF than the OC.
```{r, fig.width=6, fig.height=4}
geneList <- eg_diff_genes$logFC
names(geneList) <- eg_diff_genes$Gene
cnet <- cnetplot(egoOF_DB, foldChange = geneList, showCategory = 12) + scale_color_viridis_c(name = 'log2(FoldChange)')
cnet
```

# Wikipathways

```{r}
# system("wget https://wikipathways-data.wmcloud.org/current/gmt/wikipathways-20211110-gmt-Homo_sapiens.gmt")
wp2gene <- read.gmt('wikipathways-20211110-gmt-Homo_sapiens.gmt')
wp2gene <- wp2gene %>% tidyr::separate(term, c("name","version","wpid","org"), "%")
wpid2gene <- wp2gene %>% dplyr::select(wpid, gene) #TERM2GENE
wpid2name <- wp2gene %>% dplyr::select(wpid, name) #TERM2NAME

ewp <- enricher(eg_diff_genes$ENTREZID,
                TERM2GENE = wpid2gene,
                TERM2NAME = wpid2name,
                pvalueCutoff = 0.1)

ewp_plot <- dotplot(ewp, showCategory=10) + ggtitle("Dotplot for WikiPathways")
ewp_plot
```
## Table
```{r}
ewp <- setReadable(ewp, OrgDb =  org.Hs.eg.db, keyType = 'ENTREZID')
ewp@result %>% DT::datatable()
```

# KEGG Pathway Enrichment

```{r}
kk <- enrichKEGG(gene         = eg_diff_genes$ENTREZID, 
                 universe = eg_universe$ENTREZID,
                 organism     = 'hsa')
dotplot(kk) + ggtitle("KEGG Pathway Enrichment") 
```

## Table
```{r}
kk <- setReadable(kk, OrgDb =  org.Hs.eg.db, keyType = 'ENTREZID')
kk@result %>% DT::datatable()
```


# My takeaway(s)

1. Many many genes are different. 
2. Many relate to retina development
3. In the WikiPathways, the set relating to Neural Crest differentiation is intersting. You see MITF/DCT (melanosomes). MSX2 is a TF that is involved in craniofacial morphogenesis and limb pattern formation. This TF is downregulated in the fissure. 

# Session Info

```{r}
devtools::session_info()
```
